documentation/design_docs/design/resource-scheduler.adoc

```diff
@@ -0,0 +1,225 @@
+
+    // Check quota against limits
+    p = a.plan
+    needed = {}
+    for r in p.required_resources:
+        needed[r] = sum(usage[r])
+
+        // Only possible for 'router' resource or if only sharded queues are used
+        if usage[r].len == 1 && usage[r][0] > 1:
+            free = available[r] - usage[r][0]
+        else 
+            free = 1 - min(usage[r])
+
+        if free >= r.credit: // We can fit within existing limits
+            needed[r] += r.credit
+        else // We can't fit within existing limits, needs to add at least another pod
+            needed[r] += ceil(r.credit)
+
+        if needed[r] > allowed[r]:
+            a.status.phase = Pending // Allows it to be retried at a later point
+            a.status.message = "Quota exceeded"
+            skip to next address
+
+    if sum(needed) > allowed['aggregate']:
+        a.status.message = "Aggregate quota exceeded"
+        skip to next address
+
+    // Passed quota check, make sure we expand capacity if needed
+    // Schedule new address if needed
+    for r in p.required_resources:
+        rdef = get_resource_def(r)
+        pooled = r.credit < 1
+
+        if needed[r] > available[r]:
+            // Also creates the resource if it doesnt exist
+            // For r == router, it will be non-zero
+            // For r == broker, create with name of resource if pooled, otherwise use name of
+            // address
+            if available[r] == 0:
+                create_resource(rdef, ceil(needed[r]), pooled ? r : a.name)
+            else
+                scale_resource(rdef, ceil(needed[r]), pooled ? r : a.name)
+
+        // Special handling of pooled 'broker' resource
+        // This could be done for non-broker resources as well, it doesn't have to be taken into
+        // account by the agent. Perhaps some hint of 'scheduling policy' should be part of the
+        // resource definition?
+        if r == broker && pooled:
+            schedule_address(a, usage[r], r.credit, rdef)
+    // Mark as ready to be configured
+    a.status.phase = Configuring
+```
+
+. Scheduling for pooled addresses
++
+```
+schedule_address(address, usage, credit, rdef):
+    brokers = {}
+    // List all brokers (including newly created without addresses):
+    for b in listBrokers(rdef):
+        if usage.exists(b):
+            brokers[b] = usage[b]
+        else
+            brokers[b] = 0
+
+    // Sort by usage
+    sorted_brokers = brokers.sort_by_highest_usage()
+    for b in sorted_brokers:
+        if brokers[b] + credit < 1:
+            address.assigned_to = b
+            return
+
+    // Should never come here (error)
+
+```
+
+=== Simplified variant
+
+A variant where the usage and need calculations are the same and there is a clear distinction
+between sharded and pooled behavior throughout.
+
+. Calculate actual usage of created addresses:
++
+```
+usage = {}
+detailed_usage = {}
+for a in all addresses with phase != Pending
```

Comment:
is it all addresses in Pending, or just Active addresses?

Reply:
I see there is now a Configuring state, so it would certainly be Running + Configuring. Question I guess is whether Terminating and Failed addresses are counted.